src/middlewared/middlewared/utils/cpu.py

```diff
@@ -0,0 +1,99 @@
+import functools
+import json
+import psutil
+import re
+import subprocess
+
+from collections import defaultdict
+
+
+AMD_PREFER_TDIE = (
+    # https://github.com/torvalds/linux/blob/master/drivers/hwmon/k10temp.c#L121
+    # static const struct tctl_offset tctl_offset_table[] = {
+    'AMD Ryzen 5 1600X',
+    'AMD Ryzen 7 1700X',
+    'AMD Ryzen 7 1800X',
+    'AMD Ryzen 7 2700X',
+    'AMD Ryzen Threadripper 19',
+    'AMD Ryzen Threadripper 29',
+)
+RE_CORE = re.compile(r'^Core ([0-9]+)$')
+RE_CPU_MODEL = re.compile(r'^model name\s*:\s*(.*)', flags=re.M)
+
+
+@functools.cache
+def cpu_info() -> dict:
+    return {
+        'cpu_model': get_cpu_model(),
+        'core_count': psutil.cpu_count(logical=True),
+        'physical_core_count': psutil.cpu_count(logical=False),
+    }
+
+
+def get_cpu_model():
+    with open('/proc/cpuinfo', 'r') as f:
+        model = RE_CPU_MODEL.search(f.read())
+        return model.group(1) if model else None
+
+
+def cpu_temperatures() -> dict:
+    cp = subprocess.run(['sensors', '-j'], capture_output=True, text=True)
+    sensors = json.loads(cp.stdout)
+    amd_sensor = sensors.get('k10temp-pci-00c3')
+    if amd_sensor:
+        return _amd_cpu_temperatures(amd_sensor)
+    else:
+        return _generic_cpu_temperatures(sensors)
+
+
+def _generic_cpu_temperatures(sensors: dict) -> dict:
+    temperatures = defaultdict(dict)
+    for chip, value in sensors.items():
+        for name, temps in value.items():
+            if not (m := RE_CORE.match(name)):
+                continue
+            for temp, value in temps.items():
+                if 'input' in temp:
+                    temperatures[chip][int(m.group(1))] = value
+                    break
+
+    return {
+        str(k): v for k, v in dict(enumerate(sum(
```

Comment:
Here you return string keys in AMD implementation you return integer keys

Reply:
Good catch..fixing this asap